Test failure in mul_gen

fredrik@agm:~/src/antic$ export ANTIC_TEST_MULTIPLIER=10
    fredrik@agm:~/src/antic$ make check MOD=nf_elem
    make[1]: Entering directory '/home/fredrik/src/antic'
    make[1]: Nothing to be done for 'shared'.
    make[1]: Leaving directory '/home/fredrik/src/antic'
    make[1]: Entering directory '/home/fredrik/src/antic/nf_elem'
    get_fmpz_mod_poly....PASS
    init/clear....PASS
    set_si_ui...PASS
    mul....PASS
    set_coeff_num_fmpz....PASS
    div....PASS
    mul_gen....FAIL:
    K = Number field with defining polynomial -1/166153499473114484112975882535010304*x-1/23736214210444926301853697505001472
    a = -7
    b = 47270499
    p1 = 110297831/6148914691236517205
    p2 = -330893493
    ../Makefile.subdirs:84: recipe for target '../build/nf_elem/test/t-mul_gen_RUN' failed

Remark: the default ANTIC_TEST_MULTIPLIER=1 is really too low. Some of the tests only run 10 or 100 iterations. I would bump it at least to 10. Maybe manually add more test iterations specifically for low-degree number fields.

--- a/nf_elem/mul_gen.c
+++ b/nf_elem/mul_gen.c
@@ -22,9 +22,22 @@ void nf_elem_mul_gen(nf_elem_t a, const nf_elem_t b, const nf_t nf)
   {
       fmpz * den = LNF_ELEM_DENREF(a);
 	    fmpz * num = LNF_ELEM_NUMREF(a);
-      _fmpq_mul(num, den, LNF_ELEM_NUMREF(b), LNF_ELEM_DENREF(b), fmpq_poly_numref(nf->pol), fmpq_poly_numref(nf->pol) + 1);
+      /* _fmpq_mul assumes a positive denominator */
+      if (fmpz_sgn(fmpq_poly_numref(nf->pol) + 1) < 0)
+      {
+          fmpz_t t;
+          fmpz_init(t);
+          fmpz_neg(t, fmpq_poly_numref(nf->pol) + 1);
+          _fmpq_mul(num, den, LNF_ELEM_NUMREF(b), LNF_ELEM_DENREF(b), fmpq_poly_numref(nf->pol), t);
+          fmpz_clear(t);
+      }
+      else
+      {
+          _fmpq_mul(num, den, LNF_ELEM_NUMREF(b), LNF_ELEM_DENREF(b), fmpq_poly_numref(nf->pol), fmpq_poly_numref(nf->pol) + 1);
+          fmpz_neg(num, num);
+      }
+
       _fmpq_canonicalise(num, den);
-      fmpz_neg(num, num);
   }
   else if (nf->flag & NF_QUADRATIC)
   {

